hnn_core/lfp.py

```diff
@@ -0,0 +1,234 @@
+"""
+LFPsim - Simulation scripts to compute Local Field Potentials (LFP) from cable
+compartmental models of neurons and networks implemented in NEURON simulation
+environment.
+
+LFPsim works reliably on biophysically detailed multi-compartmental neurons
+with ion channels in some or all compartments.
+
+Last updated 12-March-2016
+Developed by :
+Harilal Parasuram & Shyam Diwakar
+Computational Neuroscience & Neurophysiology Lab, School of Biotechnology,
+Amrita University, India.
+
+Email: harilalp@am.amrita.edu; shyam@amrita.edu
+www.amrita.edu/compneuro
+
+translated to Python and modified to use use_fast_imem by Sam Neymotin
+based on mhines code
+
+References
+----------
+1 . Parasuram H, Nair B, D'Angelo E, Hines M, Naldi G, Diwakar S (2016)
+Computational Modeling of Single Neuron Extracellular Electric Potentials
+and Network Local Field Potentials using LFPsim. Front Comput Neurosci 10:65
+[PubMed].
+"""
+
+# Authors: Mainak Jas <mjas@mgh.harvard.edu>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Sam Neymotin <samnemo@gmail.com>
+
+from neuron import h
+
+import numpy as np
+from numpy.linalg import norm
+
+
+def _get_sections_on_this_rank(sec_type='Pyr'):
+    ls = h.allsec()
+    ls = [s for s in ls if sec_type in s.name()]
+    return ls
+
+
+def _get_segment_counts(all_sections):
+    '''The segment count of a section excludes the endpoints (0, 1)'''
+    seg_counts = list()
+    for sec in all_sections:
+        seg_counts.append(sec.nseg)
+    return seg_counts
+
+
+def _transfer_resistance(section, ele_pos, sigma, method):
+    """Transfer resistance between section and electrode position.
+
+    To arrive at the extracellular potential, the value returned by this
+    function is multiplied by the net transmembrane current flowing through all
+    segments of the section. Hence the term "resistance" (voltage equals
+    current times resistance).
+
+    Parameters
+    ----------
+    section : h.Section() The NEURON section. ele_pos : list (x, y, z) The x,
+        y, z coordinates of the electrode (in um) sigma : float Extracellular
+        conductivity (in S/m) method : str Approximation to use. 'psa' assigns
+        all transmembrane currents to the center point (0.5). 'lsa' treats the
+        section as a line source, but a single multiplier is calculated for
+        each section. Returns
+    -------
+    vres : list The resistance at each section.
+    """
+    ele_pos = np.array(ele_pos)  # electrode position to Numpy
+
+    start = np.array([section.x3d(0), section.y3d(0), section.z3d(0)])
+    end = np.array([section.x3d(1), section.y3d(1), section.z3d(1)])
+
+    if method == 'psa':
+
+        mid = (start + end) / 2.
+
+        # distance from section midpoint to electrode
+        dis = norm(ele_pos - mid)
+
+        # setting radius limit
+        if dis < section.diam / 2.0:
+            dis = section.diam / 2.0 + 0.1
```

Comment:
I wonder if we should throw a warning if the electrode is intersecting a neuron? However, I could see that getting annoying as the only way to fix it would be through trial and error...